src/layers/basic.jl

```
@@ -371,3 +371,55 @@ function Base.show(io::IO, m::Parallel)
   join(io, m.layers, ", ")
   print(io, ")")
 end
+
+"""
+    Embedding(in, out; init=randn)
+
+A lookup table that stores embeddings of dimension `out` 
+for a vocabulary of size `in`. 
+
+This layers is often used to store word embeddings and retrieve them using indices. 
+The input to the layer can be either a vector of indexes
+or the corresponding [onehot encoding](@ref Flux.OneHotArray). 
+
+# Examples
+
+```julia-repl
+julia> vocab_size, embed_size = 1000, 4;
+
+julia> model = Embedding(vocab_size, embed_size)
+Embedding(1000, 4)
+
+julia> vocab_idxs = [1, 722, 53, 220, 3]
+
+julia> x = OneHotMatrix(vocab_idxs, vocab_size);
+
+julia> model(x)
+4×5 Matrix{Float32}:
+  0.91139    0.670462    0.463217   0.670462    0.110932
+  0.247225  -0.0823874   0.698694  -0.0823874   0.945958
+ -0.393626  -0.590136   -0.545422  -0.590136    0.77743
+ -0.497621   0.87595    -0.870251   0.87595    -0.772696
+```
+
+julia> model(vocab_idxs) == model(x)
+true
+"""
+struct Embedding{W}
+  weight::W
+end
+
+@functor Embedding
+
+function Embedding(in::Integer, out::Integer;
+               init = (i...) -> randn(Float32, i...))
+  return Embedding(init(out, in))
+end
+
+(m::Embedding)(x::Union{OneHotVector, OneHotMatrix}) = m.weight * x # equivalent to m.weight[:,onecold(x)]
+(m::Embedding)(x::Union{Int,AbstractVector}) = m.weight[:, x]
```

Won't this run into aliasing problems in the gradient? If there are repeats in `x` then I think broadcasting will overwrite itself.

---

I think this also has the potential to limit us by not letting Base figure out the shape of the resultant indexing operation. Ideally, you wouldn't need to convert your input down to index, but rather use the dimensions to figure out the correct output dims. Thoughts?

---

> Won't this run into aliasing problems in the gradient? If there are repeats in x then I think broadcasting will overwrite itself.

I didn't know there was such problem, do we have a corresponding issue in Zygote?

---

It seems aliasing is not a problem (right?)
```julia
julia> w = [1,2,3,4];

julia> gradient(w -> sum(w[[1,1,2,2,2]]), w)
([2, 3, 0, 0],)
```

---

> I think this also has the potential to limit us by not letting Base figure out the shape of the resultant indexing operation.

julia needs to infer the type (which should figure out will be a matrix), not the shape

> Ideally, you wouldn't need to convert your input down to index, but rather use the dimensions to figure out the correct output dims. Thoughts?

I don't understand what you are suggesting, but i don't think the purpose of this layer can be achieved without indexing

---

Re aliasing, see https://github.com/FluxML/Zygote.jl/issues/821 etc.

---

> Re aliasing, see [FluxML/Zygote.jl#821](https://github.com/FluxML/Zygote.jl/issues/821) etc.

sh*t, this is a big problem for this functionality, when embedding a sentence multiple word occurrences are totally common.
So due to FluxML/Zygote.jl#821  it would be broken on gpu, even worse than that actually, it would give incorrect gradients

---

Yes. I think ScatterNNlib.jl had GPU code to avoid this, and the NNlib PR was the first step to hooking that up.

---

I see. That PR is taking quite some time, maybe I'll just implement scatter_add here for the time being.

---

> julia needs to infer the type (which should figure out will be a matrix), not the shape

Right, I meant Base specifically, and not the compiler. The `getindex` operation can also determine the resultant shape

```Julia
julia> r = reshape(1:10, 2,:)
2×5 reshape(::UnitRange{Int64}, 2, 5) with eltype Int64:
 1  3  5  7   9
 2  4  6  8  10

julia> r[rand(1:10, 3,3,3)]
3×3×3 Array{Int64,3}:
[:, :, 1] =
 5  6  7
 2  8  1
 3  1  1

[:, :, 2] =
  8  6   4
  5  5  10
 10  2   9

[:, :, 3] =
 2  7  9
 5  9  3
 4  7  9
```

---

I'm not sure I understand the relevance to the PR? Embedding layers accept indices as input. Where is the "conversion to indices" happening?

---

Where do we convert? I think I should have said reshape instead of "convert down" the shape.

---

I'm sorry, I just don't see where the reshaping or "converting down" is happening. We have a matrix whose columns are word embeddings. We want to select a subset of those embeddings. What's preventing the shape from being inferred?

---

@DhairyaLGandhi are you aware of the fact that the line of code we are commenting restricts the indexes to `Union{Int,AbstractVector}`?

---

I'm specifically trying to avoid having that type in there.

---

Oh I think I finally see what you were trying to say. You're concerned about the case where `x` > 1D if the type restrictions are lifted?

---

What's the reason for removing the hard-typing? To be more generic? I'm not sure being able to index with non-vectors is useful here, and I'm more worried about silent errors.